genestack_client/share_util.py

```diff
@@ -45,21 +74,129 @@ def share_files_for_edit(self, file_accessions, group_accession, destination_fol
+        :type destination_folder: str
+
+        :raises GenestackServerException: if some of the given files cannot be shared by the current
+        user (i.e. he doesn't own them or doesn't have the :attr:`ShareUtil.Permissions.SHARE`
+        permission).
+        """
+        permissions = self.__to_list(permissions)
+        self.__share(
+            file_accessions, group_accession, destination_folder, 'shareFiles', permissions
         )
 
-    def __share_files(self, method_name, file_accessions, group_accession, destination_folder):
-        is_iterable = isinstance(file_accessions, collections.Iterable)
-        if is_iterable and not isinstance(file_accessions, basestring):
-            file_accessions = list(file_accessions)
-        else:
-            file_accessions = [file_accessions]
+    def safe_share_files(
+            self, file_accessions, group_accession, permissions, destination_folder=None
+    ):
+        """
+        Same as :meth:`share_files` but does not throw an exception in case some of the given files
+        cannot be shared (i.e. the current user doesn't own them or doesn't have the
+        :attr:`ShareUtil.Permissions.SHARE` permission).
 
-        self.invoke(method_name, file_accessions, [group_accession])
+        :param file_accessions: accession or an iterable of accessions of files to be shared
+        :type file_accessions: str | collections.Iterable[str]
+        :param group_accession: accession of the group to share the files with
+        :type group_accession: str
+        :param permissions: permissions that should be assigned to the provided files. Must consist
+               of :class:`ShareUtil.Permissions` values
+        :type permissions: str | collections.Iterable[str]
+        :param destination_folder: accession of the folder to link shared files into. Typically
+               this parameter should be used for linking files into group folders, which is
+               currently impossible to do using the :meth:`FilesUtil.link_file` method. No links
+               will be created if this parameter is equal to `None`
+        :type destination_folder: str
+        """
+        permissions = self.__to_list(permissions)
+        self.__share(
+            file_accessions, group_accession, destination_folder, 'safeShareFiles', permissions
+        )
+
+    def __share(
+            self, file_accessions, group_accession, destination_folder, method, *args
+    ):
+        file_accessions = self.__to_list(file_accessions)
+        self.invoke(method, file_accessions, [group_accession], *args)
         if destination_folder is not None:
             self.invoke('linkFiles', file_accessions, destination_folder, group_accession)
 
+    def share_folder(self, folder_accession, group_accession, permissions, destination_folder=None):
+        """
+        Recursively share the given folder, its subfolders and files inside them.
```

Comment:
safe or not safe?

Reply:
what?

Reply:
What will happen `if some of the given files cannot be shared by the current user (i.e. he doesn't own them or doesn't have the :attr:`ShareUtil.Permissions.SHARE` permission`

Reply:
fixed